### PR TITLE
#2550 Implement downloading files via CDP

### PR DESCRIPTION
**Pull request by britka**
## Proposed changes
Add new download method **CDP**. It works similar to **FOLDER** approach but now this process is controlled by browser.
To use this download method you need to use snippet
```
$("#someId").download(DownloadOptions.using(FileDownloadMode.CDP);
```
> [!NOTE]  
> **This method works only for Chromium based browsers**

For more information see:
* [Webdriver Chrome DevTools](https://www.selenium.dev/documentation/webdriver/bidirectional/chrome_devtools/)
* [Download complete with CDP](https://www.selenium.dev/documentation/webdriver/bidirectional/chrome_devtools/cdp_api/#download-complete)
* [Chrome devtools protocol - Browser](https://chromedevtools.github.io/devtools-protocol/tot/Browser/)

## Checklist
- [*] Checkstyle and unit tests are passed locally with my changes by running `gradlew check chrome_headless firefox_headless` command
- [*] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    return archiveFile(driver, downloadedFile);
+  }
+
+  @Nullable
+  protected DownloadsFolder getDownloadsFolder(Driver driver) {
+    return driver.browserDownloadsFolder();
+  }
+
+  private void waitUntilDownloadsCompleted(long timeout, long pollingInterval) {
+    for (long start = currentTimeMillis(); currentTimeMillis() - start < timeout; pause(pollingInterval)) {
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:98]: Please use `com.codeborne.selenide.Stopwatch` for polling wait needs. See example usage in `com.codeborne.selenide.impl.SelenideElementProxy#dispatchAndRetry`

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    File archivedFile = new File(uniqueFolder, downloadedFile.getName());
+    moveFile(downloadedFile, archivedFile);
+    log.debug("Moved the downloaded file {} to {}", downloadedFile, archivedFile);
+    return archivedFile;
+  }
+
+  private void initDevTools(Driver driver) {
+    if (driver.browser().isChromium()) {
+      devTools = ((HasDevTools) driver.getWebDriver()).getDevTools();
+      devTools.createSession();
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:127]:
```suggestion
      devTools.createSessionIfThereIsNotOne();
```

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      devTools.send(Page.enable());
+    } else {
+        throw new RuntimeException("The browser that you choose doesn't have such functionality.");
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:130]: throw new RuntimeException("The browser you selected <ADD_BROWSER_NAME_HERE> doesn't have Chrome Devtools protocol functionality.");

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    }
+  }
+
+  private void prepareDownloadWithCdp(Driver driver) {
+    initDevTools(driver);
+
+    devTools.send(org.openqa.selenium.devtools.v119.browser.Browser.setDownloadBehavior(
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:137]: Please use static imports

```
import static org.openqa.selenium.devtools.v119.browser.Browser.downloadProgress;
import static org.openqa.selenium.devtools.v119.browser.Browser.downloadWillBegin;
import static org.openqa.selenium.devtools.v119.browser.Browser.setDownloadBehavior;
import static org.openqa.selenium.devtools.v119.browser.Browser.SetDownloadBehaviorBehavior;
```

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    devTools.addListener(org.openqa.selenium.devtools.v119.browser.Browser.downloadWillBegin(), handler -> {
+      fileName.set(handler.getSuggestedFilename());
+    });
+
+    devTools.addListener(
+      org.openqa.selenium.devtools.v119.browser.Browser.downloadProgress(),
+      e -> {
+        downloadComplete.set(e.getState() == COMPLETED);
+        log.info("Download is in progress");
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:151]: There appears to be a debug log message not info.

**Review by BorisOsipov — COMMENTED** [review]
Please add integration tests as in `integration.FileDownloadToFolderTest`

**Review by BorisOsipov — COMMENTED** [review]
Please add integration tests as in `integration.FileDownloadToFolderTest`

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 1 above)
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java]: Fixed

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 2 above)
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java]: Done

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 3 above)
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java]: Message has been changed

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 4 above)
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java]: Done

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 5 above)
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java]: Yes. This was a bug. I fixed it

**Comment by britka** [comment]
Add tests to ```integration.FileDownloadToFolderWithCdpTest```

**Comment by asolntsev** [comment]
@britka Few tests failed in Firefox:

```java
FileDownloadToFolderWithCdpTest > downloadLargeFile() FAILED
    java.lang.IllegalArgumentException: The browser you selected "firefox" doesn't have Chrome Devtools protocol functionality.
        at com.codeborne.selenide.impl.DownloadFileToFolderCdp.initDevTools(DownloadFileToFolderCdp.java:78)
        at com.codeborne.selenide.impl.DownloadFileToFolderCdp.prepareDownloadWithCdp(DownloadFileToFolderCdp.java:83)
        at com.codeborne.selenide.impl.DownloadFileToFolderCdp.download(DownloadFileToFolderCdp.java:49)
        at com.codeborne.selenide.commands.DownloadFile.execute(DownloadFile.java:75)
        at com.codeborne.selenide.commands.DownloadFile.execute(DownloadFile.java:29)
        at com.codeborne.selenide.commands.Commands.execute(Commands.java:161)
        at integration.FileDownloadToFolderWithCdpTest.downloadLargeFile(FileDownloadToFolderWithCdpTest.java:265)
```

You can add `assumeThat(isFirefox()).isFalse()` or something similar.

**Comment by britka** [comment]
Ok. I'll add that

**Comment by britka** [comment]
@asolntsev @BorisOsipov I will revert some changes.
In our case we should give file to user anyway. So I suggest to do the next
Variant 1.
When User chose FOLDER method we should check if it is chromium based browser then we apply CDP method
otherwise we continue using FOLDER .

Variant 2.
The same but for CDP method.

What do you think?

**Comment by BorisOsipov** [comment]
@britka 
>When User chose FOLDER method we should check if it is chromium based browser then we apply CDP method
otherwise we continue using FOLDER .

If the user has chosen FOLDER, we should use the FOLDER method and not confuse the user with some of our predictions about how we can make the file available to them.

**Comment by britka** [comment]
@BorisOsipov agree.

**Comment by britka** [comment]
@asolntsev I fixed the tests according to your comments

**Review by asolntsev — CHANGES_REQUESTED** [review]

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      if (downloadComplete.get()) {
+        log.debug("File {} download is complete", fileName.get());
+        return;
+      }
+      stopwatch.sleep(pollingInterval);
+    } while (!stopwatch.isTimeoutReached());
+  }
+
+  private void initDevTools(Driver driver) {
+    if (driver.browser().isChromium()) {
```
[the remark this answers is not in the document]
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:72]: @asolntsev I disagree with you. For example, `FirefoxDriver` is also implementing `HasDevTools` interface. But it doesn't support this functionality. It uses devtools v.85

[the remark this answers is not in the document]
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:72]: I just tried - Firefox does download the file. Only command `setDownloadBehavior` doesn't work.

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.openqa.selenium.devtools.v119.browser.model.DownloadProgress.State.COMPLETED;
+
+@ParametersAreNonnullByDefault
+public class DownloadFileToFolderCdp extends DownloadFileToFolder {
+  private static final Logger log = LoggerFactory.getLogger(DownloadFileToFolderCdp.class);
+  private AtomicBoolean downloadComplete = new AtomicBoolean(false);
```
[the remark this answers is not in the document]
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java]: Done

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:105]: I prefer to leave current implementation if you do not mind

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    devTools.addListener(Browser.downloadWillBegin(), handler -> {
+      fileName.set(handler.getSuggestedFilename());
+    });
+
+    devTools.addListener(
+      Browser.downloadProgress(),
+      e -> {
+        downloadComplete.set(e.getState() == COMPLETED);
+        log.debug("Download is in progress");
```
[the remark this answers is not in the document]
Comment (britka) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:139]: I check this in other piece of code.

File: src/test/java/com/codeborne/selenide/impl/DownloadFileToFolderCdpTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+import static com.codeborne.selenide.files.DownloadActions.click;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.UUID.randomUUID;
+import static org.apache.commons.io.FileUtils.readFileToString;
+import static org.apache.commons.io.FileUtils.writeStringToFile;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+final class DownloadFileToFolderCdpTest {
```
[the remark this answers is not in the document]
Comment (britka) [reply src/test/java/com/codeborne/selenide/impl/DownloadFileToFolderCdpTest.java]: I deleted them

File: src/test/java/com/codeborne/selenide/impl/DownloadFileToFolderCdpTest.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  void setUp() {
+    when(linkWithHref.driver()).thenReturn(driver);
+    when(linkWithHref.findAndAssertElementIsInteractable()).thenReturn(link);
+    when(linkWithHref.toString()).thenReturn("<a href='report.pdf'>report</a>");
+  }
+
+  private static WebDriver setUpDriver() {
+    ChromeOptions chromeOptions = new ChromeOptions();
+    chromeOptions.addArguments("--headless=new");
+    return new ChromeDriver(chromeOptions);
```
[the remark this answers is not in the document]
Comment (britka) [reply src/test/java/com/codeborne/selenide/impl/DownloadFileToFolderCdpTest.java]: Now this not actual. I've deleted unit tests for this class.

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:139]: Yes, but this log is lying. It says `Download is in progress`, but the download is actually completed.

File: src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java:105]: I don't understand why we forbid this functionality in Firefox if it works.

**Review by asolntsev — APPROVED** [review]
LGTM

@britka Checkstyle warning: 
> Unused import at FileDownloadToFolderWithCdpTest.java:14:8

**Comment by britka** [comment]
> LGTM
> 
> @britka Checkstyle warning:
> 
> > Unused import at FileDownloadToFolderWithCdpTest.java:14:8

@asolntsev Fixed unused imports

**Pull request merged**